modules/plan.nix

```diff
@@ -29,10 +29,25 @@ in
   # combined or replaced. We seed the package Options with an empty set forcing the
   # default values.
   options = {
-    packages = mkOption {
-      type = attrsOf package;
+    plan-json = mkOption {
+      type = attrsOf unspecified;
     };
-
+    use-package-keys = mkOption {
+      type = bool;
+      default = false;
+    };
+    package-keys = mkOption {
+      type = listOf str;
+      default = [];
+    };
+    packages = if !config.use-package-keys
```

Comment:
> Why do we need `addPackageKeys` everywhere?

I had to change the type of `packages` to make `pkgs.lib.modules.mkAliasDefinitions` work here:
https://github.com/input-output-hk/haskell.nix/blob/a4795893f301e9dc4df8b595b41e62e89d66bd1f/modules/install-plan/override-package-by-name.nix#L8

This means we need to know the list of all the keys for `packages` when creating the type.

If you add an override for a package in the plan it should work fine, because we include every package id and name in `package-keys` here:
https://github.com/input-output-hk/haskell.nix/blob/a4795893f301e9dc4df8b595b41e62e89d66bd1f/modules/install-plan/override-package-by-name.nix#L5).

If you try to override a package that is not in the plan it will now raise an error.  This might actually be quite a nice feature.

However we often want to include overrides that only apply to some plans (for instance packages only included for a given platform).  In those cases we can either try to duplicate the logic that determines if the package will be in the plan and only include the override when it is, or we can add it to `package-keys`.

To make that second option a little easier, `addPackageKeys` takes the `attrNames` for `packages` and includes them in `package-keys` (it can only do this for `packages` and not `config.packages` to avoid infinite recursion).

Reply:
> it might lead to recursion issues that will be hard to debug

It scares me too, but I think it might be fairly safe.  Do you have a concrete example of something that might fail?

I'm not too worried about `use-package-keys` since it is not something anyone should need to use directly (perhaps we could call it `is-cabal-project` or something).

`package-keys` is like a way of forward declaring the packages you are going to override.  One thing does not work due to infinite recursion is getting `addPackageKeys` from the module arguments in any way.

```nix
{modules = [({haskellLib, ...}: haskellLib.addPackageKeys { package.x.flags.y = false; })];}
```

However that is not really related to the `addPackageKeys` function.  You get the same infinite recursion issue if you reference any of the module args at the top level.  It even happens if you try to look up a non existent function, for instance this fails with the same `error: infinite recursion encountered`:

```nix
{modules = [({haskellLib, ...}: haskellLib.foo {})];}
```

These alternatives all work:

```nix
{modules = [(haskellLib.addPackageKeys {packages.x.flags.y = false;})];} # Where haskellLib is in scope
{modules = [(rec {package-keys = builtins.attrNames packages; packages.x.flags.y = false;})];}
{modules = [((x: x // {package-keys = builtins.attrNames x.packages;}) {packages.x.flags.y = false;})];}
```